File: javascript/courseSchedule.js
var canFinish = function(n, prerequisites) {
  const adj = {};
  const indegree = {};
  const bfs = [];
  
  for (let i = 0; i < n; i++) {
    adj[i] = [];
  }
  
  for (let pair of prerequisites) {
    const [course, prereq] = pair;
    adj[prereq].push(course);
    if (!indegree[course]) indegree[course] = 0;
    indegree[course]++;
  }
  
  console.log('adj', adj);
  console.log('indegree', indegree);
  
  for (let i = 0; i < n; i++) {
    if (!indegree[i]) bfs.push(i);
  }

  console.log('bfs', bfs);

  for (let node of bfs) {
    for (let i = 0; i < adj[node].length; i++) {
      const j = adj[node][i];
      if (--indegree[j] === 0) bfs.push(j);
      console.log('bfs', bfs);
    }
  }
  return Object.keys(bfs).length === n;
};

// const canFinish = function(numCourses, prerequisites) {
//   const dependencies = new Array(numCourses);
//   const addDependencies = (course, dependency) => {
//     if (!dependencies[course]) {
//       dependencies[course] = {};
//     }

//     if (!dependencies[course][dependency]) {
//       dependencies[course][dependency] = true;
//       if (dependencies[dependency]) {
//         const arr = Object.keys(dependencies[dependency]);
//         for (let el of arr) {
//           addDependencies(course, el);
//         }
//       }
//     }
//     return;
//   }
  
//   for (let prereq of prerequisites) {
//     addDependencies(prereq[0], prereq[1]);
//   }

//   return dependencies;
// };

console.log(canFinish(2, [[1,0]]));
console.log(canFinish(3, [[1,0],[2,1]]));

File: python/course_schedule.py
from collections import defaultdict

class Solution(object):
    def canFinish(self, n, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        g = [[] for i in range(n)]
        degree = [0] * n

        for course, prereq in prerequisites:
          g[prereq].append(course)
          degree[course] += 1

        bfs = [i for i in range(n) if degree[i] == 0]

        for i in bfs:
          for j in g[i]:
            degree[j] -= 1
            if degree[j] == 0:
              bfs.append(j)
        
        return len(bfs) == n

s = Solution()
print(s.canFinish(2, [[1,0]]))
print(s.canFinish(2, [[1,0],[0,1]]))